### DISCUSSION TITLE
Call Graph on ticket received

### DISCUSSION THREAD

**Discussion by xperiandri** [AzureAD/microsoft-identity-web#1773] [Q&A]
I wanted to gather additional AAD user info and merge it into principal claims.
Because UserInfo endpoint is not being called (https://github.com/AzureAD/microsoft-identity-web/pull/1771) and I need some more info (https://github.com/AzureAD/microsoft-identity-web/discussions/1772) decided to use Graph. 

So I provided a delegate to `events.OnTicketReceived` property.
But as long as a user is not authenticated there yet, Graph authentication fails.
Moreover, it cannot find an authentication schema. As long as I override it to use Cookie as default.
``` C#
.AddMicrosoftIdentityWebApp(
	options => configuration.GetSection("AzureAd").Bind(options),
	options =>
	{
		configuration.GetSection("AzureAd").Bind(options);
		if (!options.ResponseType.Split(' ').Contains("code"))
		{
			options.ResponseType = string.Join(' ', "code", options.ResponseType);
		}
		ExternalClaimsMappings.mapMicrosoftClaims(options.ClaimActions);
		var events = options.Events;
		//events.OnTokenResponseReceived = MicrosoftAuthenticationEventsHandler.OnTokenResponseReceived;
		//events.OnTicketReceived = MicrosoftAuthenticationEventsHandler.OnTicketReceived;
	},
	openIdConnectScheme: IdentityProviders.Microsoft,
	displayName: "Microsoft Account",
	cookieScheme: null)
```

So I provided a delegate to `OnTokenResponseReceived` property to do
``` F#
[<CompiledName("OnTokenResponseReceived")>]
let onTokenResponseReceived (ctx : TokenResponseReceivedContext) = task {
    let provider =
        DelegateAuthenticationProvider(
            fun request ->
                request.Headers.Authorization <-
                    System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", ctx.ProtocolMessage.AccessToken)
                Task.CompletedTask)
    let graph = GraphServiceClient(provider)
    let principal = ctx.Principal |> MicrosoftPrincipal.op_Implicit
    let! userInfo = graph.GetUserInfoAsync (principal, CancellationToken.None)
    let identity = ctx.Principal.Identity :?> ClaimsIdentity
    identity.AddClaim (Claim(TdlClaims.Email, userInfo.Email))
    // TODO: Add other claims
}
```
but AccessToken is null because of https://github.com/AzureAD/microsoft-identity-web/issues/880 which I fixed in https://github.com/AzureAD/microsoft-identity-web/pull/1771

So my question is how to instruct Graph to take the token received before signing in completed in order to gather user info and merge it into claims?

**Comment by jmprieur** [comment]
@xperiandri : you can use the following sample as an example on how to call Graph from the OnTokenValidated event:

https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/3918633c09c8f90275901939ad3844ab65a5ebb7/5-WebApp-AuthZ/5-2-Groups/Startup.cs#L51

cc: @kalyankrishna1 to help if needed.

**Reply by xperiandri** [reply]
@jmprieur this works great unless I add a new scope.
In that case on login, I get `UiRequiredException` even though the user has already given consent and logged in to AAD.
How to fix that?

**Reply by jmprieur** [reply]
UiRequiredException is expected, @xperiandri: this means the user needs to consent for more.
If you need more scopes at the beginning, you need to specify them in the EnableTokenAcquisitionToCallDownstreamWebApi()

**Comment by jmprieur** [comment]
@xperiandri : can you pre-authorize this scope (admin consent)?

**Reply by xperiandri** [reply]
What I want to achieve:
1. Auth server used for login. Regular users must have `UserRead` and `TeamWork.ReadAll`
2. For Teams marketplace page administrator must have `Organization.ReadAll` in order to fill info required to create an account in our system.
3. After the admin created an account in our system I want him to give consent for whole organization for `UserRead` and `TeamWork.ReadAll` in order not to distract users with that. As users are pupils of schools.

So my current setup is:
* ASP.NET Core app
* Razor Class Library project with Blazor Server hosted on a separate area inside the app above

My issues:
1. I want to have 2 sets of initial scopes: for regular users and for admin
2. I want to authenticate admin redirected from Teams Marketplace with Azure AD scheme only. While users coming to other areas of the app can use any provider or login and password. In this setup, Blazor Server auth does not work at all. Neither when I set Azure AD scheme nor when I use authorization without a particular scheme. And token acquisition always fails. Whatever I pass to it or Graph service client.

**Reply by xperiandri** [reply]
1. So how can I use 2 different sets of scopes on login?

**Reply by xperiandri** [reply]
2. How to fix the issue with Blazor Server?

**Reply by xperiandri** [reply]
3. Why whole Microsoft.Identity.Web is tied to a scheme named `OpenIdConnect`? After I renamed my Azure AD scheme to this name I got a constant loop of `ConsentHandler.HandleException(ex);` -> `/MicrosoftAccount/Identity/Challenge/` and back

**Reply by jmprieur** [reply]
@xperiandri : Microsoft.Identity.Web uses the default scheme, which is the scheme passed to the last AddAuthentication("scheme"). To use several schemes, use the overrides of AddMicrosoftIdentityWebApp that take the authentication scheme. and use the right authentication scheme to call Microsoft Graph. 

See https://github.com/AzureAD/microsoft-identity-web/wiki/Multiple-Authentication-Schemes

**Reply by xperiandri** [reply]
This is my setup
``` C#
services.ConfigureApplicationCookie(options =>
{
    options.LoginPath = "/Identity/Account/Login";
    options.LogoutPath = "/Identity/Account/LogOff";
});

// Configure Identity to use the same JWT claims as OpenIddict instead
// of the legacy WS-Federation claims it uses by default (ClaimTypes),
// which saves you from doing the mapping in your authorization controller.
services.Configure<IdentityOptions>(options =>
{
    options.ClaimsIdentity.UserIdClaimType = Claims.UserId;
    options.ClaimsIdentity.UserNameClaimType = Claims.Name;
    options.ClaimsIdentity.EmailClaimType = Claims.Email;
    options.ClaimsIdentity.RoleClaimType = Claims.Role;
});
services.AddTransient<IClaimsTransformation, TdlClaimsTransformation>();

services.Configure<MicrosoftIdentityOptions>(IdentityProviders.Microsoft, configuration.GetSection("AzureAd"));

services
// I need standard ASP.NET Identity
    .AddAuthentication(CookieAuthenticationDefaults.AuthenticationScheme)
    .AddMicrosoftWorkOrSchool(configuration);
```

``` C#
public static AuthenticationBuilder AddMicrosoftWorkOrSchool(this AuthenticationBuilder builder, IConfiguration configuration)
{
    var initialScopes = configuration["DownstreamApi:Scopes"]?.Split(' ') ?? configuration["MicrosoftGraph:Scopes"]?.Split(' ');

    builder
        .AddMicrosoftIdentityWebApp(
            options =>
            {
                configuration.GetSection("AzureAd").Bind(options);
// This is never called because you throw away the Access and call to UserInfo endpoint does not happen
                //ExternalClaimsMappings.mapMicrosoftClaims(options.ClaimActions);
            },
            openIdConnectScheme: IdentityProviders.Microsoft, // This is my scheme 'MicrosoftAccount'
            displayName: "Microsoft Account",
            cookieScheme: null)
        .EnableTokenAcquisitionToCallDownstreamApi(initialScopes)
        .AddMicrosoftGraph(configuration.GetSection("MicrosoftGraph"))
        .AddDownstreamWebApi("DownstreamApi", configuration.GetSection("DownstreamApi"))
        .AddInMemoryTokenCaches();

    return builder;
}
```
Could you tell me what is wrong with my setup?

**Comment by xperiandri** [comment]
How can I change login claims set on the fly?

``` C#
builder
    .AddMicrosoftIdentityWebApp(
        options =>
        {
            var events = options.Events;
            events.OnRedirectToIdentityProvider = ctx =>
            {
                if (ctx.HttpContext.Request.RouteValues["area"] == TDL.Model.Common.Constants.Areas.Marketplace)
                {
                    var options = ctx.Options;
                    var scope = options.Scope;
// This manipulation does not work
                    scope.Remove("Team.ReadBasic.All");
                    scope.Add("Directory.Read.All");
                }
                return Task.CompletedTask;
            };
            events.OnTokenValidated = MicrosoftAuthenticationEventsHandler.OnTokenValidated;
        },
        openIdConnectScheme: IdentityProviders.Microsoft,
        displayName: "Microsoft Account",
        cookieScheme: null)
```
How can I achieve that?
